sshconnect2.c

```diff
@@ -1262,10 +1262,20 @@ identity_sign(struct identity *id, u_char **sigp, size_t *lenp,
 			goto out;
 		}
 		sign_key = prv;
-		if (sshkey_is_sk(sign_key)) {
+		if (sshkey_is_sk(sign_key) &&
+		    (sign_key->sk_flags & SSH_SK_USER_PRESENCE_REQD)) {
+			fmprintf(stdout, "You may need to touch your "
+				 "authenticator.\n");
+		}
+	}
+ retry_pin:
+	if ((r = sshkey_sign(sign_key, sigp, lenp, data, datalen,
+	    alg, options.sk_provider, pin, compat)) != 0) {
+		debug_fr(r, "sshkey_sign");
+		if (pin == NULL && !retried && sshkey_is_sk(sign_key) &&
+		    r == SSH_ERR_KEY_WRONG_PASSPHRASE) {
```

Comment:
Changing the PIN prompting logic from "inspect key flags, prompt, try to sign" to "try to sign, look at error, prompt for PIN if needed, retry" is a step in the right direction, IMHO. It will also allow us to kill SSH_SK_USER_VERIFICATION_REQD.

Reply:
Thanks, I think so too.  And thanks for reviewing the patches!